coffee/shepherd.coffee

```diff
@@ -117,6 +117,13 @@ class Step extends Evented
     @tether = new Tether extend(tetherOpts, @options.tetherOptions)
 
   show: =>
+    if @options && @options.beforeShowPromise && typeof @options.beforeShowPromise is 'function'
```

Comment:
This is basically what I was thinking. A couple quick CoffeeScript tweaks:

- `@options` is guaranteed to be defined to at least an empty object `{}` thanks to [the constructor](https://github.com/HubSpot/shepherd/blob/master/coffee/shepherd.coffee#L45)
- The existential operator `?` does more than just loosely translate to `x !== null || x !== undefined`. When using it on a function call, say `@options.beforeShowPromise?()`, it will check that `typeof fun is 'function'`. CoffeeScript.org [has a nice repl](http://coffeescript.org/#try:x%3F.fun()%0A%0Ax.fun%3F()) you can play with.

One way to go about it would be to check that the result of `@options.beforeShowPromise?()` is defined (which you can use `?` for), and then either attach the promise or call the function. It may make things a bit clearer by defining the result anyways.